rollup/da_syncer/da_syncer.go

```diff
@@ -14,39 +15,66 @@ var (
 )
 
 type DASyncer struct {
+	l2EndBlock uint64
 	blockchain *core.BlockChain
 }
 
-func NewDASyncer(blockchain *core.BlockChain) *DASyncer {
+func NewDASyncer(blockchain *core.BlockChain, l2EndBlock uint64) *DASyncer {
 	return &DASyncer{
+		l2EndBlock: l2EndBlock,
 		blockchain: blockchain,
 	}
 }
 
 // SyncOneBlock receives a PartialBlock, makes sure it's the next block in the chain, executes it and inserts it to the blockchain.
-func (s *DASyncer) SyncOneBlock(block *da.PartialBlock) error {
+func (s *DASyncer) SyncOneBlock(block *da.PartialBlock, override bool, sign bool) error {
 	currentBlock := s.blockchain.CurrentBlock()
 
 	// we expect blocks to be consecutive. block.PartialHeader.Number == parentBlock.Number+1.
-	if block.PartialHeader.Number <= currentBlock.Number().Uint64() {
+	// if override is true, we allow blocks to be lower than the current block number and replace the blocks.
+	if !override && block.PartialHeader.Number <= currentBlock.Number().Uint64() {
 		log.Debug("block number is too low", "block number", block.PartialHeader.Number, "parent block number", currentBlock.Number().Uint64())
 		return ErrBlockTooLow
 	} else if block.PartialHeader.Number > currentBlock.Number().Uint64()+1 {
 		log.Debug("block number is too high", "block number", block.PartialHeader.Number, "parent block number", currentBlock.Number().Uint64())
 		return ErrBlockTooHigh
 	}
 
-	parentBlock := s.blockchain.GetBlockByNumber(currentBlock.Number().Uint64())
+	parentBlockNumber := currentBlock.Number().Uint64()
+	if override {
+		parentBlockNumber = block.PartialHeader.Number - 1
+		// reset the chain head to the parent block so that the new block can be inserted as part of the new canonical chain.
+		err := s.blockchain.SetHead(parentBlockNumber)
```

Comment:
Any concern that the removed blocks will be re-propagated to us and we'll overwrite the chain? Or that's not the case because we disabled the network protocols?

Reply:
Yeah we disabled the network protocols (this is just an extension of the L1 follower mode basically). So we operate in isolation. This is either needed as a preparation step 
- for the sequencer to recover from permissionless mode 
- a user who wants to submit batches in permissionless mode

Reply:
But if there is a long unfinalized chain of L2 blocks, then after we recover from enforced batch mode and re-enable L2 sync, it's possible that this old chain will overwrite the new chain, right?